2020-01-Winter/1_linked_lists/README.md

```diff
@@ -140,19 +141,76 @@ Go to [Top](#top)
 
 ### 3. SOLUTION 3 TODO :bug:
 
-Source: TODO :bug:
+Source: Cracking the Code Interview Book, Chapter 2, Question 2.1
 
-#### Naive/Simple Solution 
+#### Solution 
 
-TODO :bug:
+With the Buffer: Use a set to keep track of what integers have been encountered.
 
-#### Optimal Solution
+```
+void RemoveDuplicatesWithBuffer(Node *Start) {
+  if (Start == nullptr || Start->GetNext() == nullptr) {
+    return;
+  }
 
-TODO :bug:
+  unordered_set<int> set;
+
+  Node *Curr = Start;
+  Node *Prev = nullptr;
+
+  while (Curr != nullptr) {
+    // Doesn't exist
+    if (set.find(Curr->GetValue()) == set.end()) {
+
+      set.insert(Curr->GetValue());
+      Prev = Curr;
+      Curr = Curr->GetNext();
+
+    } else { // Exists
+
+      Prev->SetNext(Curr->GetNext());
+      delete Curr;
+      Curr = Prev->GetNext();
+    }
+  }
+}
+```
+
+Without a Buffer: Have a pointer/runner that iterates over the list to check for duplicates.
+
+```
+void RemoveDuplicatesWithoutBuffer(Node *Start) {
+  if (Start == nullptr || Start->GetNext() == nullptr) {
+    return;
+  }
+
+  Node *Prev = Start;
+  Node *Curr = Prev->GetNext();
+
+  while (Curr != nullptr) {
+
+    Node *Check = Start;
+
+    while (Check != Curr) { // Check for earlier duplicates
+
+      if (Check->GetValue() == Curr->GetValue()) {
+        Prev->SetNext(Curr->GetNext());
+        Curr = Curr->GetNext();
+        break; // All other duplicates have been removed
+      }
+      Check = Check->GetNext();
+    }
```

Comment:
Nitpick: I think that checking prior nodes is perfectly valid; keep in mind that you could also check later nodes if you want to preserve order of appearance of unique elements. In some situations this might matter, but for our purposes, this is fine.

Reply:
My original solution would check future nodes as well. The reason I changed it to this implementation was that I was having memory leaks in the other version and the textbook suggested this solution instead.
Honestly, looking back at it now, iterating over the entire list seems like a better idea...